### PR TITLE
Message ordering and accepted ranges for infrastructure

### PR DESCRIPTION
**Pull request by akoshelev**
## Overview

This change makes infrastructure layer to order messages it sends down to the network layer. `RecordId` is used to order messages and every batch regardless whether it is complete batch or no will have messages sent in order. 

In order to keep the order and prevent buffers from grow infinitely large, a concept of currently accepted range is used. It is tightly coupled to the batch size used inside the infrastructure layer. For instance, if batch size is 10, then infra starts with accepting send requests for records 0 to 9. These send requests do not need to be send in order, but until all records from 0 to 9 are received, infra blocks 

Note that this implementation simply rejects sends requests for records greater than the maximum accepted value, but we will revisit it. The ideal strategy is to slow down (block) rather than simply reject, but it would be easy to implement once we're trained to write unit tests for our protocols with the presence of accepted ranges of records.

Most of the changes in unit tests I had to make were simple - just set initial record id to 0 and that's it. However there is one protocol `Attribution` that takes advantage of parallel multiplications and does it in a way where it is not deterministic how many multiplications can be spawned in parallel (from 1 to 4). Setting batch size to 4 is correct but makes the execution really slow, so I had to add another parameter to reduce the time to flush to 1ms.

### CODE REVIEW

File: src/helpers/buffers.rs [hunk 1]
```diff
@@ -131,3 +194,231 @@ impl ReceiveBuffer {
...
+                // don't drop the original vector
+                let mut data = ManuallyDrop::new(value.data);
+
+                // reinterpret Vec<Option<T>> as Vec<T>
+                Vec::from_raw_parts(
+                    data.as_mut_ptr().cast::<MessageEnvelope>(),
+                    data.len(),
+                    data.capacity(),
+                )
+            }
```
Comment (benjaminsavage) [inline src/helpers/buffers.rs:277]: Could this potentially change in some future version of Rust? I mean, the memory layout? Doesn't this approach kind of bind us to an implementation detail?

File: src/protocol/securemul.rs [hunk 2]
```diff
@@ -88,134 +88,6 @@ impl<'a, N: Network> SecureMul<'a, N> {
 #[derive(Error, Debug)]
 pub enum Error {}
 
-/// Module to support streaming interface for secure multiplication
```
Comment (benjaminsavage) [inline src/protocol/securemul.rs:91]: Why did you remove this whole streaming multiplication thing?

File: src/helpers/buffers.rs (hunk 1 above)
Comment (akoshelev) [reply src/helpers/buffers.rs:277]: I don't think so because it is documented behaviour and in many places in standard library it is expected for `Option<T>` to have the same layout as `T`. Changing this would likely be a major breaking change so I don't think they will even consider that

File: src/protocol/securemul.rs (hunk 2 above)
Comment (akoshelev) [reply src/protocol/securemul.rs:91]: ah yea, sorry I mentioned this in the commit, but did not mention in the PR description. It is basically useless behaviour for us because that is not how we use multiplications. Cost of maintaining it raises - for example there was a hidden bug here: `stream_element_idx` was moved inside the closure so it was never incremented more than once. I tried to fix it but there is no easy way to do that (maybe `Arc<AtomicU32>>` could be a fix). 

Given that we don't observe this pattern, this is dead code and I think it should be removed, but we could reconsider that if there is an argument for keeping it around

File: src/helpers/buffers.rs (hunk 1 above)
Comment (martinthomson) [reply src/helpers/buffers.rs:277]: I don't think that that is strictly true.  `Option<T>` can require a different layout than `T`, though for some types it can share a layout (and that is the general preference where possible).  For instance, `Option<u32>` cannot take just 4 bytes as it has $2^32+1$ different possible values, though `Option<&T>` as a reference type can always use a null value to indicate `None`.

File: src/helpers/buffers.rs [hunk 3]
```diff
@@ -3,13 +3,58 @@ use crate::protocol::RecordId;
...
+    /// Once every element in the vec is set, it is considered full and data can be moved
+    /// away from it. `SendBatch` guarantees no extra allocations in this case because this vector
+    /// is just reinterpreted as `Vec<MessageEnvelope>`.
+    ///
+    /// If `data` is not full and buffer needs to be flushed to network, a new vector will be
+    /// allocated and `data` will be copied element-by-element which is less efficient and therefore
+    /// should be avoided if possible
+    ///
+    /// This implementation guarantees that this vector is sorted according to `MessageEnvelope::record_id`
+    data: Vec<Option<MessageEnvelope>>,
```
[the remark this answers is not in the document]
Comment (martinthomson) [reply src/helpers/buffers.rs:49]: I think that this could also use a type of `[Option<Box<[u8]>>; N]` where `N` is twice the batch size.  If everything is working properly, you will want to be able to receive messages that are out of order by some number.  Ideally, that is a small number, at least smaller than the batch size.

But you want to be able to eject a slice from this store for sending when it is ready, while you are also receiving new messages to send in the next batch.  That suggests that you need a store that is large enough for a whole batch, plus whatever your reordering tolerance is.

But you also want this store to be a clean multiple of the batch size so that you don't have to worry about weird offsets into the store.  (Maybe you can do that for that exceptional case you talk about in the comment, but it's OK if that costs a lot.)

An alternative is to maintain two buffers, b1 and b2, with each maintaining an initial offset (or with a single starting offset and a flag saying which one is primary, or you can mem::swap pointers to them, I don't care).

[the remark this answers is not in the document]
Comment (martinthomson) [reply src/helpers/buffers.rs:49]: Another way of approaching this is to change the message so that each item is fixed length.  Then, you can store a byte array of fixed size, with an associated bit vector (BitVec from the bit_vec crate perhaps?) that indicates which values are yet to be populated.  Then, writing to the network only involves handing off a reference to the slice of bytes.

I might prefer that approach best in fact.  It's much simpler than your current design.  It also doesn't really pay a huge tax for the case where you need to write out some of the values.

[the remark this answers is not in the document]
Comment (martinthomson) [reply src/helpers/buffers.rs:49]: Note that you can move the buffer size into the type definition using const generics.  I don't know if that is worthwhile, but you might consider it if moving to a message type that is fixed size.

[the remark this answers is not in the document]
Comment (akoshelev) [reply src/helpers/buffers.rs:49]: I like the fixed length approach. Do you have a suggestion for how to pick the appropriate message length? we seem to send messages that carry either 1 or 2 field values, but I was assuming that we may send a whole batch of values (permutation)? but even to decide between 1 and 2, we would need to either allocate space for the largest possible message (50% overhead for 1 field value messages) or somehow deliver this metadata to the buffer definition

We could use const generic parameter for `Step` but generic  expressions are not a thing yet so we can't simply say `payload: [u8; Step::MESSAGE_SIZE]`...

[the remark this answers is not in the document]
Comment (martinthomson) [reply src/helpers/buffers.rs:49]: Yeah, I don't think that we need to have the buffer be declared as `[u8; N]`; we can probably just use `Vec<u8>` and allocate it at runtime, right at the start.  For now, we can just have the message type implement a trait that has a const for its encoded length (in bits eventually perhaps, but start with bytes) and has an `write_to(&self, &mut [u8])` or `write_to(&self, buf: &mut [u8], offset: usize)` method or something like that.

[the remark this answers is not in the document]
Comment (akoshelev) [reply src/helpers/buffers.rs:49]: Fixed size messages appears to be working (I picked 8 bytes for message size for now and use in-place buffers) in my prototype. I also realized that writing out just some values from the buffer wouldn't work without extra bookkeeping: we can't really move the window until *all* messages have been received. It may be possible to keep the bit vec around until all records have been received, but I don't see a clear benefit for it as buffer capacity is not increased anyway. 

Wondering if I should just remove this logic to write out partial buffers and try to mitigate the issues with `Attribution` by keeping buffers that are multiples of `batch_size` and multiplier is > 1

**Comment by akoshelev** [comment]
Please ignore the changes until this PR is no longer in WIP

**Comment by akoshelev** [comment]
this is a draft, please don't review it

**Comment by benjaminsavage** [comment]
I think this diff is now pretty out of date. In fact, I think most of these changes have already landed. Can we abandon this PR now?
